projects/vdk-plugins/vdk-notebook/src/vdk/plugin/notebook/notebook.py

```diff
@@ -0,0 +1,120 @@
+
+log = logging.getLogger(__name__)
+
+
+class JobNotebookLocator:
+    """
+    Locate the data job files that would be executed by us.
+    """
+
+    @staticmethod
+    def get_notebook_files(directory: pathlib.Path) -> List[pathlib.Path]:
+        """Locates the files in a directory, that are supported for execution.
+        Files supported for execution are: .ipynb
+        Other files in the directory are ignored.
+        :return: List of files from the directory that supported for execution, sorted alphabetically by name.
+        :rtype: :class:`.list`
+        """
+        script_files = [
+            x for x in directory.iterdir() if (x.name.lower().endswith(".ipynb"))
+        ]
+        script_files.sort(key=lambda x: x.name)
+        log.debug(f"Script files of {directory} are {script_files}")
+        return script_files
+
+
+class Notebook:
+    """
+    Given a notebook file locates the cells with "vdk" tag and saves them.
+    Files supported for reading are: ipynb.
+    Other cells are ignored.
+    """
+
+    def __init__(self, file_path: Path):
+        try:
+            self.sql_and_run_cells = []
+            self.python_helper_cells = []
+            self.file_path = file_path
+
+            content = json.loads(file_path.read_text())
+            cells = [
+                cell
+                for cell in content["cells"]
+                if cell["cell_type"] == "code"
+                and "vdk" in cell["metadata"].get("tags", {})
+            ]
+            for cell in cells:
+                code = "".join(cell["source"])
+                if code.startswith("%sql") or "def run(" in code:
+                    if code.startswith("%sql"):
+                        code = "".join(cell["source"])
+                        code.replace(";", "")
+                        self.sql_and_run_cells.append(
+                            [code.replace("%sql", ""), TYPE_SQL]
+                        )
+                    else:
+                        self.sql_and_run_cells.append(
+                            ["".join(cell["source"]), TYPE_PYTHON]
+                        )
+                else:
+                    self.python_helper_cells.append("".join(cell["source"]))
+
+        except json.JSONDecodeError as e:
+            errors.log_and_rethrow(
+                to_be_fixed_by=errors.ResolvableBy.USER_ERROR,
+                log=log,
+                what_happened=f"Failed to read the {file_path.name} file.",
+                why_it_happened=f"The provided {file_path.name} cannot be loaded into json format and "
+                f"cannot be read as a Jupyter notebook",
+                consequences=errors.MSG_CONSEQUENCE_TERMINATING_APP,
+                countermeasures=f"Check the {file_path.name} format again",
+                exception=e,
+                wrap_in_vdk_error=True,
+            )
+
+    def extract_notebook_steps(self, context: JobContext):
+        for index, cell in enumerate(self.sql_and_run_cells):
```

Comment:
nit: since by design `self.sql_and_run_cells` populated (during init) is a requirement for this method implementation, you may assert on it for the purpose of fail-fast. That makes the future feature extending support less error-prone.